add BiasAddGradient

vnghia opened the pull request:
@saxenasaurabh 

This draft PR shows you how I want to split the test. In summary, we have some several files for testing right now. 
- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/eager/gradients_test.cc ( for Math gradients )
- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/eager/mnist_gradients_test.cc
- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/eager/gradient_checker_test.cc ( for NN gradients )

And utils:
- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/eager/gradients_util.h
- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/eager/unified_api_testutil.h
- https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/eager/mnist_gradients_testutil.h

Since the gradients are classed by `math` and `nn`, but the tests aren't, it causes some overlap. Like `AddGradModel` is written twice.
https://github.com/tensorflow/tensorflow/blob/6c2616c1ea7d580546d9ac818a70e6edf40c56f4/tensorflow/c/eager/mnist_gradients_testutil.h#L40
https://github.com/tensorflow/tensorflow/blob/6c2616c1ea7d580546d9ac818a70e6edf40c56f4/tensorflow/c/eager/gradients_test.cc#L77


Furthermore, `gradients_util.h` and `unified_api_testutil.h` has some overlaps too
https://github.com/tensorflow/tensorflow/blob/6c2616c1ea7d580546d9ac818a70e6edf40c56f4/tensorflow/c/eager/gradients_util.h#L85
https://github.com/tensorflow/tensorflow/blob/6c2616c1ea7d580546d9ac818a70e6edf40c56f4/tensorflow/c/eager/unified_api_testutil.h#L49

https://github.com/tensorflow/tensorflow/blob/6c2616c1ea7d580546d9ac818a70e6edf40c56f4/tensorflow/c/eager/gradients_util.h#L43-L46
https://github.com/tensorflow/tensorflow/blob/6c2616c1ea7d580546d9ac818a70e6edf40c56f4/tensorflow/c/eager/unified_api_testutil.h#L56-L58

---

I propose that we should:
- `*GradModel` should be in `math_grad_utils.h` and `nn_grad_utils.h`
- `Test*Grad` should be in `math_grad_test.h` and `nn_grad_test.h`
- Merge `gradients_util.h` and `unified_api_testutil.h`
- Have an unified process for testing the gradient, because currently, some gradients are tested with `gradient_checker` and some are tested by writing the result manually.

Part of #42668

vnghia committed:
add BiasAddGradient

vnvo2409 commented:
@saxenasaurabh 
Could you take a look at this PR ? Thank you !

saxenasaurabh commented on tensorflow/c/experimental/gradients/nn_grad_testutil.cc:
```diff
@@ -0,0 +1,71 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
```
Could we define `BiasAddGradModel` in `nn_grad_test.cc` and not have `nn_grad_testutil.*`?

vnvo2409 replied:
Done

saxenasaurabh commented on tensorflow/c/experimental/gradients/nn_grad_test.cc:
```diff
@@ -0,0 +1,154 @@
+#include "tensorflow/c/eager/c_api_unified_experimental_internal.h"
+#include "tensorflow/c/eager/gradients.h"
+#include "tensorflow/c/eager/gradients_internal.h"
+#include "tensorflow/c/eager/gradients_util.h"
```
Let's try to use `unified_api_testutil.h` instead of `gradients_util.h` here. We can slowly replace all usages of `gradients_util.h` with `unified_api_testutil.h`.

vnvo2409 replied:
Done

saxenasaurabh reviewed (commented):
The duplication of `*GradModel` probably indicates that we have duplicate tests. I think we should be able to move op-level tests in `mnist_gradients_test` to respective {math|nn|array}_grad_test.cc and get rid of the `<Op>GradModel` in `mnist_gradients_testutil`.

> gradients_util.h and unified_api_testutil.h has some overlaps too

Let's try to use `unified_api_testutil.h` going forward. We can slowly delete `gradients_util.h`.

> *GradModel should be in math_grad_utils.h and nn_grad_utils.h

Keeping these models in util files might not be necessary. We could keep them in the respective test.cc files.

I like the idea of having `module_grad_test.cc` files, let's do that.

> Have an unified process for testing the gradient

Yep, using gradient checker would be nice and would greatly simplify adding new tests.

vnghia committed:
Use gradient checker for testing BiasAddGrad

vnvo2409 commented:
In e5cf403, I add `grad_test_helper` which is a helper function to automatically test the model against `gradient_checker`. One thing I don't understand, why `gradient_checker` modifies the `inputs`. I think it should not touch the `inputs` at all. I added a TODO for revisiting this point. We could merge this PR as-is, and I will send some follow-up PRs for clean up the test.

vnghia committed:
cleanup `nn_grad_test.cc`

Changes:

- add `CompareWithGradientsCheckers` to automatically test the model against `gradient_checker`.

- `TF_MODEL_FACTORY` and `TF_GRAD_MODEL_FACTORY` will expand to models that are needed for testing.

- Move `RegisterGradients` and `BuildImmediateExecutionContext` to `Setup`.

vnvo2409 commented:
@saxenasaurabh 
Could you take a look at this PR ? Thank you!

vnghia committed:
restore `inputs` in `gradients_checker`

vnvo2409 commented:
> One thing I don't understand, why `gradient_checker` modifies the `inputs`.

Fixed in 68b0bd3

vnghia committed:
Move `TF_MODEL_FACTORY` and `TF_GRAD_MODEL_FACTORY` to a seperate header

saxenasaurabh commented on tensorflow/c/experimental/gradients/model_factory_helper.h:
```diff
@@ -0,0 +1,98 @@
+#ifndef TENSORFLOW_C_EXPERIMENTAL_GRADIENTS_MODEL_FACTORY_MACRO_H_
+#define TENSORFLOW_C_EXPERIMENTAL_GRADIENTS_MODEL_FACTORY_MACRO_H_
+
+// This macro will expand to a function that defines a `Model`. This `Model` is
```
Let's not add these macros right now. We need to fuse the `Model` and `RunModel` impls in `gradients_util.h` and `unified_api_testutil.h` and update the gradient checker to use the non-registry version of `Model` as well.

vnvo2409 replied:
How about a PR for merging `gradients_util.h` and `unified_api_testutil.h`. After that PR is merged, we could rebase it onto this PR which is more cleaner.

vnvo2409 replied:
After some investigations, I think we better do the fusion in a separate PR. `gradient_checker_test` now uses the `Model` defined in `mnist_gradients_testutil.h`. If we want to update the `gradient_checker` to use the non-registry version of `Model`, we have to update `mnist_gradients_testutil.h`( and a lot of files which are dependent on `mnist_gradients_testutil.h` ) and there will be a lot of changes.

I propose we follow this process:
- First, we merge this PR, with the macros.
- Next, I will make a PR for updating these macros to non-registry version along with updating `gradient_checker` and rewriting completely `gradient_checker_test` ( it has overlaps right now ) with the new version of these macros. We can safely remove `mnist_gradients_testutil.h` and `gradient_utils.h` out of `gradient_checker`.
- Next, a PR for moving the tests from `gradients_test` and `mnist_gradients_test` to `module_grad_test`. while moving to `module_grad_test`, we could update the test to use the non-registry version of `Model` and `CompareWithGradientsCheckers`. In addition, we could safely delete `gradients_test` ( not `mnist_gradients_test` because there are some tests for MNIST ).
- Finally, we remove `gradient_utils` and `mnist_gradients_testutil`, update the MNIST tests in `mnist_gradients_test` and maybe move everything related to gradients to `tensorflow/c/experimental/gradients` ( so we could have the same directory structure as python ).

The cleanup process could take a while but I think it's worth doing. What do you think ?

saxenasaurabh replied:
I am working on refactoring the `Tape` API that would greatly simplify the model functions. Here's a draft [proposal](https://github.com/tensorflow/community/compare/master...saxenasaurabh:cpp-gradients?expand=1&short_path=c3fb7c7#diff-c3fb7c75a838948b459de9b64e6edc8d2f75b49fdf7dad8040f6724f00fb85f8) that I will publish as a formal RFC in a day or so (let me know if you have any suggestions!). I have the code mostly ready for that. Once that is done the macro would look like:

```
#define TF_GRAD_MODEL_FACTORY(name, num_inputs, num_outputs, num_grad_outputs, \
                              ops_call)                                        \
  Status name(AbstractContext* ctx,                                            \
              absl::Span<AbstractTensorHandle* const> inputs,                  \
              absl::Span<AbstractTensorHandle*> outputs,                       \
              const GradientRegistry& registry) {                              \
    std::unique_ptr<Tape>(new Tape(/*persistent=*/false));                                \
    for (int i{}; i < num_inputs; ++i) {                                       \
      tape->Watch(inputs[i]);                                            \
    }                                                                          \
                                                                               \
    AbstractTensorHandle* temp_outputs[num_outputs] = {};                      \
    AbstractContextPtr tape_ctx(new TapeContext(ctx, tape, registry));         \
    ops_call;                                                                  \
    TF_RETURN_IF_ERROR(tape->ComputeGradient(                                  \
        ctx, temp_outputs, inputs,                          \
        /*output_gradients=*/{}, outputs));                                 \
                                                                               \
    for (int i{}; i < num_outputs; ++i) {                                      \
      temp_outputs[i]->Unref();                                                \
    }                                                                          \
    return Status::OK();                                                       \
  }
```

If we want we could further add helpers to watch a list of tensors. Also we need to figure out how to avoid manually `Unref`ing intermediate tensors. I have an idea for this that might require a larger refactor. Once all this is done the model fn will be a fairly short. 

Another problem with this macro style is that you rely on variables created in the macro in your `ops_call`. That does not seem like a great pattern to me.

My suggestion for now would be to drop the macros and use the verbose model function. Then think about how to de-dup the `Model` and `RunModel` and move existing tests to `module_grad_test`. That will be a great cleanup I think. Meanwhile I will try to check in the new API. Once that is done we can discuss what additional helpers we need to make the tests less verbose.

vnvo2409 replied:
Done

saxenasaurabh commented on tensorflow/c/experimental/gradients/model_factory_helper.h:
```diff
@@ -0,0 +1,98 @@
+              absl::Span<AbstractTensorHandle* const> inputs,          \
+              absl::Span<AbstractTensorHandle*> outputs,               \
+              const GradientRegistry& registry) {                      \
+    auto tape = new Tape(/*persistent=*/false);                        \
```
Since we do not wish to compute gradients here, this function is effectively just calling `ops_call`.

vnvo2409 replied:
Agree. However, it saves us a lot of copy paste ( like new Tape, Watch, ... ) so I think this macro might be useful. In addition, a model without computing gradients is needed for `gradient_checker`.

saxenasaurabh replied:
Since we never call `tape->ComputeGradient` here, `tape->Watch`, `TapeContext` etc are no-ops. So this macro is effectively:

```
#define TF_MODEL_FACTORY(name, num_inputs, num_outputs, ops_call)      \
  Status name(AbstractContext* ctx,                                    \
              absl::Span<AbstractTensorHandle* const> inputs,          \
              absl::Span<AbstractTensorHandle*> outputs,               \
              const GradientRegistry& registry) {                      \
    ops_call;                                                          \
    return Status::OK();                                               \
}
```

vnvo2409 replied:
Since the old models use `tape->ComputeGradient`, `tape->Watch`, etc, I just copy without thinking too much. If it is this much simple, we don't need a macro.

vnvo2409 commented on tensorflow/c/eager/gradient_checker.cc:
```diff
@@ -99,8 +99,8 @@ Status RunAndMaybeSum(AbstractContext* ctx, Model forward,
-  TF_RETURN_IF_ERROR(ops::Sum(ctx, absl::MakeSpan(sum_inputs),
-                              absl::MakeSpan(model_outputs), "sum_output"));
+  TF_RETURN_IF_ERROR(
+      ops::Sum(ctx, sum_inputs, absl::MakeSpan(model_outputs), "sum_output"));
```
No. `absl::MakeSpan` returns a modifiable span of `sum_inputs`. But `ops::Sum` only needs a `absl::Span<AbstractTensorHandle* const> inputs` which could not be modified.

vnvo2409 commented on tensorflow/c/eager/gradient_checker.cc:
```diff
@@ -191,6 +191,8 @@ Status CalcNumericalGrad(AbstractContext* ctx, Model forward,
   }
 
+  // Restore the inputs
+  inputs[input_index] = theta;
```
> It may be cleaner to create a shallow copy of inputs that we can update instead of updating in-place and restoring.

I've already though about it. However, I could not find a way to create a shallow copy. We have to create a `std::vector<AbstractTensorHandle*>` because another span still modifies the input. In addtion, with `absl::Span<AbstractTensorHandle* const> inputs`, I could not create a `std::vector<AbstractTensorHandle*>` because there is a loss of constness ( `abseil` seems to be very strict with this behavior )

saxenasaurabh replied:
Hmm, that's interesting. So something like

```
Status CalcNumericalGrad(AbstractContext* ctx, Model forward,
                         absl::Span<AbstractTensorHandle* const> inputs,
                         int input_index, bool use_function,
                         AbstractTensorHandle** numerical_grad) {
  std::vector<AbstractTensorHandle*> inputs_copy(inputs.size());
  for (int i = 0; i < inputs.size(); i++) {
    inputs_copy[i] = inputs[i];
  }
```

doesn't work?

vnvo2409 replied:
It works but it seems sub-optimal 😕

vnvo2409 commented on tensorflow/c/experimental/gradients/grad_test_helper.cc:
```diff
@@ -0,0 +1,76 @@
+
+void CompareWithGradientsCheckers(Model model, Model grad_model,
+                                  AbstractContext* ctx,
+                                  absl::Span<AbstractTensorHandle*> inputs,
```
See previous comment.

vnvo2409 commented on tensorflow/c/eager/gradient_checker.h:
```diff
@@ -43,6 +43,9 @@ namespace gradients {
  * `numerical_grad` is the pointer to the AbstractTensorHandle* which will
  * hold the numerical gradient data at the end of the function.
+ *
+ * Note that this function will modify `inputs` and restore it to the original
```
See previous comment.

vnghia committed:
remove MODEL macro

vnghia committed:
`gradient_checker` takes const inputs

vnvo2409 commented:
One more thing, should we move gradient-related files into `tensorflow/c/experimental/gradients` ?

saxenasaurabh commented on tensorflow/c/experimental/gradients/grad_test_helper.cc:
```diff
@@ -0,0 +1,75 @@
+namespace gradients {
+namespace internal {
+
+void CompareWithGradientsCheckers(
```
Maybe use a more informative name such as `CompareNumericalAndAutodiffGradients`.

vnvo2409 replied:
Done

saxenasaurabh commented on tensorflow/c/experimental/gradients/grad_test_helper.cc:
```diff
@@ -0,0 +1,75 @@
+           TF_TensorByteSize(analytical_tensor));
+
+    for (int j = 0; j < num_elem_numerical; j++) {
+      ASSERT_NEAR(dnumerical[j], danalytical[j], /*abs_error=*/1e-2);
```
Maybe make `abs_error` be an optional argument of this function?

vnvo2409 replied:
Done

saxenasaurabh reviewed (commented):
Thanks, this looks great now! A few minor comments.

> One more thing, should we move gradient-related files into tensorflow/c/experimental/gradients ?
Yeah, we should. I am doing a bunch of refactoring for some other APIs. So I can move the gradients as well.

vnghia committed:
rename `CompareWithGradientChecker` to `CompareNumericalAndAutodiffGradients`

vnvo2409 commented:
FYI, The tests failed mainly because of `llvm`.

saxenasaurabh commented:
> FYI, The tests failed mainly because of `llvm`.

Yep llvm broke windows. A fix has been submitted I think. 
There is a pending change to fix the macos build as well.

vnvo2409 commented:
> Yep llvm broke windows. A fix has been submitted I think.
There is a pending change to fix the macos build as well.

I think we could rebase master if you need.

saxenasaurabh commented on tensorflow/c/experimental/gradients/nn_grad.cc:
```diff
@@ -110,6 +111,48 @@ class SparseSoftmaxCrossEntropyWithLogitsGradientFunction
+
+    // Recover data format from forward pass for gradient.
+    std::string data_format;
+    forward_attrs.Get("data_format", &data_format);
```
This fails internally. Change this to `TF_RETURN_IF_ERROR(forward_attrs.Get("data_format", &data_format));`

vnvo2409 replied:
Done

saxenasaurabh commented on tensorflow/c/experimental/gradients/BUILD:
```diff
@@ -95,3 +111,38 @@ filegroup(
+    srcs = [
+        "nn_grad_test.cc",
+    ],
+    args = ["--heap_check=local"],
```
Also drop these args.

vnvo2409 replied:
Done

vnghia committed:
fix internal build

saxenasaurabh commented on tensorflow/c/experimental/gradients/nn_grad_test.cc:
```diff
@@ -0,0 +1,137 @@
+};
+
+TEST_P(CppGradients, TestBiasAddGrad) {
+  if (std::get<0>(GetParam()) == "mlir" && !std::get<2>(GetParam())) {
```
`std::get<0>(GetParam()) == "mlir"` also complains. Change to `strcmp(std::get<0>(GetParam()), "mlir") == 0 `

Ideally we should use the pattern in unified_api_test: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/eager/unified_api_test.cc;l=55-58;drc=0dd94c4ad3046f4e383b371c60f272e72f9d2525

vnvo2409 replied:
Done

vnghia committed:
use `UseFunction` and `UseMlir`

saxenasaurabh commented on tensorflow/c/experimental/gradients/nn_grad_test.cc:
```diff
@@ -0,0 +1,141 @@
+    UnifiedCAPI, CppGradients,
+    ::testing::Combine(::testing::Values("graphdef", "mlir"),
+                       /*tfrt*/ ::testing::Values(false),
+                       /*executing_eagerly*/ ::testing::Values(true, false)));
```
s/executing_eagerly/use_function

vnvo2409 replied:
Done

saxenasaurabh commented on tensorflow/c/experimental/gradients/nn_grad_test.cc:
```diff
@@ -0,0 +1,141 @@
+
+  ASSERT_NO_FATAL_FAILURE(CompareNumericalAndAutodiffGradients(
+      BiasAddModel, BiasAddGradModel, ctx_.get(), {A.get(), Bias.get()},
+      /*use_function=*/!std::get<2>(GetParam()), registry_));
```
Change to `/*use_function=*/UseFunction()`

vnvo2409 replied:
Done

saxenasaurabh commented on tensorflow/c/experimental/gradients/nn_grad_test.cc:
```diff
@@ -0,0 +1,141 @@
+};
+
+TEST_P(CppGradients, TestBiasAddGrad) {
+  if (!UseFunction() && UseMlir()) {
```
This should be `if (UseFunction() && UseMlir()) {`

vnvo2409 replied:
It doesn't work. It seems that `UseFunction()` should be `!std::get<2>(GetParam())` but I followed the same implementation as in `unified_api_test`
https://github.com/tensorflow/tensorflow/blob/4679dcf1eed1d13bdc3f6288afcd46395b8e84c0/tensorflow/c/eager/unified_api_test.cc#L38

saxenasaurabh replied:
See my other comment

> Change to `/*use_function=*/UseFunction()`

vnvo2409 replied:
I changed it to `bool UseFunction() const { return !std::get<2>(GetParam()); }`

saxenasaurabh commented on tensorflow/c/experimental/gradients/nn_grad_test.cc:
```diff
@@ -0,0 +1,141 @@
+    UnifiedCAPI, CppGradients,
+    ::testing::Combine(::testing::Values("graphdef", "mlir"),
+                       /*tfrt*/ ::testing::Values(false),
+                       /*executing_eagerly*/ ::testing::Values(true, false)));
```
s/executing_eagerly/use_function

vnvo2409 replied:
Done

vnghia committed:
use `UseFunction()` for `CompareNumericalAndAutodiffGradients`

saxenasaurabh commented on tensorflow/c/experimental/gradients/nn_grad_test.cc:
```diff
@@ -0,0 +1,141 @@
+
+ public:
+  bool UseMlir() const { return strcmp(std::get<0>(GetParam()), "mlir") == 0; }
+  bool UseFunction() const { return !std::get<2>(GetParam()); }
```
I think this was fine the way it was. `bool UseFunction() const { return std::get<2>(GetParam()); } `

vnvo2409 replied:
You was right. It should be `if (UseFunction() && UseMlir()) {`. I did not change `/*use_function=*/!std::get<2>(GetParam()), registry_));` to `/*use_function=*/UseFunction()` so it caused an error. Sorry for the back and forth.

saxenasaurabh replied:
No worries. Looks all good now. Let's try submitting again.

vnghia committed:
Revert `UseFunction()` to original

saxenasaurabh commented on tensorflow/c/experimental/gradients/BUILD:
```diff
@@ -1,5 +1,21 @@
 load("//tensorflow/core/platform:rules_cc.bzl", "cc_library")
+load(
+    "//tensorflow:tensorflow.bzl",
+    "if_libtpu",
```
Loaded symbol "if_libtpu" is unused. Please remove it.

vnvo2409 replied:
Done

saxenasaurabh commented on tensorflow/c/experimental/gradients/BUILD:
```diff
@@ -1,5 +1,21 @@
+load(
+    "//tensorflow:tensorflow.bzl",
+    "if_libtpu",
+    "tf_cc_test",
```
Loaded symbol "tf_cc_test" is unused. Please remove it.

vnvo2409 replied:
Done

saxenasaurabh commented on tensorflow/c/experimental/gradients/BUILD:
```diff
@@ -1,5 +1,21 @@
+    "//tensorflow:tensorflow.bzl",
+    "if_libtpu",
+    "tf_cc_test",
+    "tf_copts",
```
Loaded symbol "tf_copts" is unused. Please remove it.

vnvo2409 replied:
Done

saxenasaurabh commented on tensorflow/c/experimental/gradients/BUILD:
```diff
@@ -1,5 +1,21 @@
+    "tf_cc_test",
+    "tf_copts",
+    "tf_cuda_cc_test",
+    "tf_cuda_library",
```
Loaded symbol "tf_cuda_library" is unused. Please remove it.

vnvo2409 replied:
Done

vnghia committed:
Remove unused in bazel

saxenasaurabh commented:
Looks like we need to rebase.

vnvo2409 commented:
> Looks like we need to rebase.

Done

vnvo2409 commented:
The internal checks failed again 😢

saxenasaurabh commented:
> The internal checks failed again 😢

Looks like a memory leak in nn_grad_test. Are you able to run your test with `-fsanitize=address`? I will patch your change and run internally as well.

vnghia committed:
Disable cuda asan for nn_grad_test

vnvo2409 commented on tensorflow/c/experimental/gradients/BUILD:
```diff
@@ -95,3 +109,37 @@ filegroup(
+        "nn_grad_test.cc",
+    ],
+    linkstatic = tf_kernel_tests_linkstatic(),
+    tags = tf_cuda_tests_tags(),
```
So the leaks are comming from `gradient_checker` and not `nn_grad_test` itself ?

saxenasaurabh replied:
Yep, seems so.

saxenasaurabh commented on tensorflow/c/experimental/gradients/BUILD:
```diff
@@ -95,3 +109,37 @@ filegroup(
+    size = "small",
+    srcs = [
+        "nn_grad_test.cc",
+    ],
```
We need to add `    args = ["--heap_check=local"],` back.

That will suppress the memory leak. I will file a bug to clean that up.

vnvo2409 replied:
Done

vnghia committed:
add `--heap_check=local`

vnvo2409 commented:
Everything seems good this time 🎉

saxenasaurabh commented:
> Everything seems good this time 🎉

Let's maybe start tracking what kind of things pass in opensource but fail internally. Ideally there should be no difference.

vnvo2409 commented:
> Let's maybe start tracking what kind of things pass in opensource but fail internally. Ideally there should be no difference.

I got this problem a lot. I think mainly because the internal CI is somewhat stricter. Tracking is a good idea but I don't have any experience. Do you have any suggestions ?
